sale_order_carrier_auto_assign/models/sale_order.py

```diff
@@ -0,0 +1,31 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def action_confirm(self):
+        for rec in self:
+            if not rec.company_id.carrier_auto_assign:
+                continue
+            rec._add_delivery_carrier_on_confirmation()
+        return super().action_confirm()
+
+    def _add_delivery_carrier_on_confirmation(self):
+        """Automatically add delivery.carrier on sale order confirmation"""
+        for order in self:
```

Comment:
I see `action_confirm` has iterated over the records in sale.order and called `_add_delivery_carrier_on_confirmation` function for each individual record. So why do we need to iterate over it again, you can consider it with function that directly handles that record and call `ensure_one` function will make the code clearer